Tenders and contracts lists pages for Dashboards module

eltercero opened the pull request:
Closes #2997

## :v: What does this PR do?

Integrates the markup done is #3028 with the remote csv data defined in Gobierto Dashboard settings. This PR enables several new pages for Contracts and Tenders dashboard in the main site:

- Summary
- Contracts index 
- Contract details
- Tenders index
- Tender details

Some parts, like the filters and the summary, are empty on purpose since this is still a work in progress.

Navigation should work between main parts (Summary ,Tenders and Contracts) and within the details page of every contract/tender. 

## :mag: How should this be manually tested?

It's available in staging: https://burjassot.gobify.net/dashboards/contratos/resumen

## :eyes: Screenshots

![Captura de pantalla 2020-05-11 a las 12 09 32](https://user-images.githubusercontent.com/545235/81550323-55096f00-9380-11ea-9118-4213fe475cb8.png)

![Captura de pantalla 2020-05-11 a las 12 09 21](https://user-images.githubusercontent.com/545235/81550339-589cf600-9380-11ea-9c47-5ed3a18b7002.png)

![Captura de pantalla 2020-05-11 a las 12 09 42](https://user-images.githubusercontent.com/545235/81550351-5cc91380-9380-11ea-94ca-f2bcd79e01b9.png)

## :shipit: Does this PR changes any configuration file?

- [ ] new environment variable in `.env.example`?
- [ ] new entry in `config/application.yml`?
- [ ] new entry in `config/secrets.yml`?

(Changes in these files might need to update the role in Ansible)

## :book: Does this PR require updating the documentation?

- [ ] new site configuration variable?
- [ ] new site template?
- [ ] new module/submodule settings?
- [ ] significant changes in some feature?

eltercero committed:
Several changes in routes and controller actions

- Renaming routes to the Spanish equivalents
- Removing "tenders" route since it's actually inside "contracts".
- Removing index route because it won't be neccessary

eltercero committed:
Views and controller with the proper info from the module settings

eltercero committed:
Vue integration: Base containers and functional navigation

eltercero committed:
Initial table and table row views integration for contracts and tenders

eltercero committed:
Integration tests for contracts and tenders lists.

eltercero committed:
Adding a test that checks that the dashboards pages can't be accessed if disabled.

eltercero committed:
Removing missing console.log

ferblape reviewed (changes requested):
Hey! It looks great, just a few details (they can be fixed here or open a new issue if you prefer):

- [x] format the prices with locale separators and units. You can check how it's been implemeted in Mataro investments

- [x] if I'm in a contract or a tender, I'd expect the navigation option to be active:

![Screenshot 2020-05-12 at 10 33 22](https://user-images.githubusercontent.com/17616/81661208-12a56800-943c-11ea-8a6a-66bba53b3b2c.png)


-  [x] data should be sorted by date ASC. You can't trust it will be sorted in the CSV
- [ ] in contracts item page add the disccount %. Anyway I'm creating a new issue with more columns we could display to make it more interesting

Crashillo commented on app/assets/stylesheets/module-dashboards.scss:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  &-home-nav {
+    background-color: rgba($color_headers, 0.04);
```
Isn't there any CSS custom properties about?

Crashillo commented on app/assets/stylesheets/module-dashboards.scss:
```diff
@@ -0,0 +1,212 @@
+      justify-content: flex-end;
+    }
+
+    ul {
```
Avoid to tag HTML elements. Use class instead

Crashillo commented on app/assets/stylesheets/module-dashboards.scss:
```diff
@@ -0,0 +1,212 @@
+      }
+    }
+
+    &--tr:hover + &--tr > * > * {
```
Uff. Pain.

Crashillo commented on app/assets/stylesheets/module-dashboards.scss:
```diff
@@ -0,0 +1,212 @@
+@import "css-conf";
```
If you've copy&paste this file, let us know where comes from to identify the legacy elements, as SASS variables, in the original files. You'll see that on next comments

eltercero replied:
@Crashillo yes, I copy-pasted this file from `module-investments.scss`. I think I did not add any style of my own, I just chose what I thought were the relevant styles.

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -1,3 +1,87 @@
+import Vue from "vue";
+import VueRouter from "vue-router";
+import * as d3 from 'd3'
```
Two things here:
1. `d3` may not be imported as a whole component (feel free to ask if you don't know how)
2. Unused in this file, so remove it

Crashillo replied:
I'm still seeing `d3` here

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -1,3 +1,87 @@
+Vue.use(VueRouter);
+Vue.config.productionTip = false;
+
 export class ContractsController {
```
@ferblape this code is actually the same as data or investments. So, shouldn't we create some creator pattern or so, in order to remove similar code? We ask 'cause we don't know how this module works

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -1,3 +1,87 @@
+        const remoteCsvData = {
+          contractsData: fullData[0],
+          tendersData: fullData[1]
+        };
```
Destructuring https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

eltercero replied:
thanks 😄

Crashillo commented on app/javascript/gobierto_dashboards/webapp/components/Table.vue:
```diff
@@ -0,0 +1,69 @@
+    };
+  },
+  created() {
+    this.labelEmpty = I18n.t("gobierto_dashboards.dashboards.contracts.empty_table");
```
Move to data

Crashillo commented on app/javascript/gobierto_dashboards/webapp/components/Table.vue:
```diff
@@ -0,0 +1,69 @@
+    <table class="dashboards-home-main--table">
+      <thead>
+        <th
+          v-for="column in columns"
```
Destructuring.
```suggestion
          v-for="{ translation } in columns"
```

Crashillo commented on app/javascript/gobierto_dashboards/webapp/components/Table.vue:
```diff
@@ -0,0 +1,69 @@
+        mode="out-in"
+      >
+        <TableRow
+          v-for="item in items"
```
Same as above

eltercero replied:
I later use `item` and `item.id` for the child component. Can I use destructuring to get both the object and one of the attributes?

Crashillo replied:
Humm, I see what happens. Let it go in the meantime, there's something I wanna explain you tomorrow (v-slot). This is a good use case.

Regarding your question, you cannot. You'd have to fetch `item` as usual. Use destructuring whenever you can use it, because it's cleaner and avoid DOM errors, but is not always available.

eltercero replied:
Yeah, it's much better. Thanks again!

Crashillo commented on app/javascript/gobierto_dashboards/webapp/components/Table.vue:
```diff
@@ -0,0 +1,69 @@
+  methods: {
+    showAll(isAllVisible) {
+      this.isAllVisible = isAllVisible
+    }
```
Unused

Crashillo commented on app/javascript/gobierto_dashboards/webapp/components/TableRow.vue:
```diff
@@ -0,0 +1,35 @@
+<template>
+  <tr class="dashboards-home-main--tr" @click="navigateTo(item)" >
+    <td
+      v-for="column in columns"
```
Destructuring

Crashillo commented on app/javascript/gobierto_dashboards/webapp/components/TableRow.vue:
```diff
@@ -0,0 +1,35 @@
+<template>
+  <tr class="dashboards-home-main--tr" @click="navigateTo(item)" >
```
Instead of adding an event for clicking, we should use the vue-router directive `router-link`, you may set the HTML tag to be rendered also. Take a look at the docs

Crashillo commented on app/javascript/gobierto_dashboards/webapp/components/TableRow.vue:
```diff
@@ -0,0 +1,35 @@
+  methods: {
+    navigateTo(item) {
+      this.$router.push({ name: this.routingMember, params: { id: item.id, item } });
+    },
```
Doing the vue-router stuff, then this method becomes useless

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/contract/ContractsIndex.vue:
```diff
@@ -0,0 +1,37 @@
+  },
+  data() {
+    return {
+      contractsData: this.$root.$data.contractsData,
```
What's that?

Crashillo replied:
Why this stuff is not coming via properties?

eltercero replied:
Ok, understood!  I'll change it in the latest PR to avoid conflict between three branches.

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/contract/ContractsIndex.vue:
```diff
@@ -0,0 +1,37 @@
+</template>
+
+<script>
+import ContractsShow from "./ContractsShow.vue";
```
Unused in DOM

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/contract/ContractsIndex.vue:
```diff
@@ -0,0 +1,37 @@
+        {field: 'contractor', translation: I18n.t('gobierto_dashboards.dashboards.contracts.contractor')},
+        {field: 'final_amount', translation: I18n.t('gobierto_dashboards.dashboards.contracts.final_amount')},
+        {field: 'end_date', translation: I18n.t('gobierto_dashboards.dashboards.contracts.end_date')},
+      ]
```
When you declare element in the `data` function, vue creates observers to their members, so that practice is discouraged. If you need such complex structure:

- You may declare it in the created function, if you would need to alter it or you would need any library precharged in the DOM previously (i.e. `I18n`).
- Since the array content looks like static, perhaps it would be better option to get them out of the Vue component, to be declared beneath the imports declarations

eltercero replied:
You're right, thanks.

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/contract/ContractsShow.vue:
```diff
@@ -0,0 +1,83 @@
+    this.labelTenderAmount = I18n.t('gobierto_dashboards.dashboards.contracts.tender_amount')
+    this.labelContractAmount = I18n.t('gobierto_dashboards.dashboards.contracts.contract_amount')
+    this.labelStatus = I18n.t('gobierto_dashboards.dashboards.contracts.status')
+    this.labelProcessType = I18n.t('gobierto_dashboards.dashboards.contracts.process_type')
```
Move to data

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/contract/ContractsShow.vue:
```diff
@@ -0,0 +1,83 @@
+    }
+  },
+  updated() {
+    this.$emit("active-tab", 1);
```
Jarl!? Every single time any vue property is changed, you emit an event? What for?

eltercero replied:
I meant `created()`, but it's no use anyway. Changed it.

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/contract/ContractsShow.vue:
```diff
@@ -0,0 +1,83 @@
+    this.labelProcessType = I18n.t('gobierto_dashboards.dashboards.contracts.process_type')
+
+    const { item } = this.$route.params;
+    const itemId = this.$route.params.id;
```
```suggestion
    const { item, id: itemId } = this.$route.params;
```

Crashillo replied:
What is `item`? is an object? An object in the url? O_O

eltercero replied:
I (wrongly) assumed vue router worked in a different manner. I'll fix it.

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/contract/ContractsShow.vue:
```diff
@@ -0,0 +1,83 @@
+        return contract.id === itemId
+      });
+    }
+  }
```
Related with the previous comment. Depending on what item is, may this clause to be useless?

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Aside.vue:
```diff
@@ -0,0 +1,14 @@
+<template>
+  <div class="pure-u-1 pure-u-lg-1-4">
+    <aside class="investments-home-aside--gap">
```
Be aware of copy&pasting. This class could be never loaded

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Nav.vue:
```diff
@@ -0,0 +1,76 @@
+  created() {
+    this.labelSummary = I18n.t("gobierto_dashboards.dashboards.contracts.summary");
+    this.labelContracts = I18n.t("gobierto_dashboards.dashboards.contracts.contracts");
+    this.labelTenders = I18n.t("gobierto_dashboards.dashboards.contracts.tenders");
```
You know where ;)

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Nav.vue:
```diff
@@ -0,0 +1,76 @@
+    },
+    tabIndexFromRouteName(name){
+      return this.$options.routesMapping.indexOf(name);
+    }
```
Won't all this stuff do what vue-router is supposed to do?

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/tender/TendersIndex.vue:
```diff
@@ -0,0 +1,36 @@
+<template>
```
This file is EXACTLY the same as ContractsIndex. That means if should be turn into a shared component, right? Pass them their properties to make them different

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/tender/TendersShow.vue:
```diff
@@ -0,0 +1,69 @@
+<template>
```
Same component as ContractsShow. Same errors explained above. Same components explanation as TendersIndex.

Crashillo commented on app/javascript/gobierto_dashboards/webapp/mixins/get_remote_data.js:
```diff
@@ -0,0 +1,13 @@
+import * as d3 from 'd3'
```
Look at the gobierto_data factories. This must be similar.

- Don't fetch `d3`. Use axios instead
- Declare here the endpoints

Crashillo replied:
Also move the file inside a `lib` folder to distinguish from other stuff

Crashillo commented on app/javascript/gobierto_dashboards/webapp/mixins/store.js:
```diff
@@ -0,0 +1,8 @@
+export const store = {
```
@ferblape Rethink if we should redeclare all this kind of things by module

eltercero committed:
Format money strings with correct punctuation and currency.

Crashillo commented on app/views/gobierto_dashboards/dashboards/contracts.html.erb:
```diff
@@ -1 +1,23 @@
+       data-logo-url="<%= @site.configuration.logo_with_fallback %>"
+       data-home-url="<%= gobierto_dashboards_contracts_url %>"
+       data-contracts-endpoint="<%= @dashboards_config.dig('data_urls', 'contracts') %>"
+       data-tenders-endpoint="<%= @dashboards_config.dig('data_urls', 'tenders') %>" z>
```
@ferblape is this the best way to share this kind of info? Shouldn't be think in some different way?

ferblape replied:
typo: `z>`

Crashillo commented on config/locales/gobierto_dashboards/views/ca.yml:
```diff
@@ -1,7 +1,27 @@
+          del %{entity_name}
+        empty_table: No hi ha dades disponibles
+        end_date: Data
+        final_amount: Amount
```
Quantitat

Crashillo commented on config/locales/gobierto_dashboards/views/ca.yml:
```diff
@@ -1,7 +1,27 @@
   gobierto_dashboards:
+    dashboards:
+      contracts:
+        assignee: Assignee
```
Adjudicatari

Crashillo commented on config/locales/gobierto_dashboards/views/en.yml:
```diff
@@ -1,7 +1,27 @@
+        contract_amount: Contract Amount
+        contractor: Contractor
+        contracts: Contracts
+        description: Visualizations and analisys for the awarded contracts and tenders
```
Analysis

Crashillo reviewed (changes requested):
There are two CSV files commited, is that right?

eltercero committed:
Removing unused method

eltercero committed:
Removing unused component

eltercero committed:
Wrong class name due to copy paste

eltercero committed:
Locale files changes

eltercero committed:
Removing unused component

eltercero committed:
Sorting contracts and tenders by date ASC

eltercero committed:
Always search for the item based on the id.

eltercero committed:
Moving i18n string initialization from created() hook to data().

eltercero committed:
Destructuring attributes in v-for

eltercero committed:
Moving columns configuration to its own file

And removing them from data()

eltercero committed:
Moving get_remote_data inside /lib

eltercero committed:
Fix navigation bar so it highlights the current tab in details pages (contract show and tender show)

eltercero commented:
@ferblape since there is a new issue for the new columns, I'll do the last thing there if that's ok! The other things you mentioned are now fixed. 

Routing and nav are not great atm though. I'll redo them after I talk with @Crashillo.

eltercero committed:
Adding format to tender show as well

Fixing specs too, since the default order changed and now everything is properly formatted.

ferblape commented on app/javascript/gobierto_dashboards/webapp/lib/config.js:
```diff
@@ -0,0 +1,12 @@
+export const contractsColumns = [
```
Should we move the columns to the module configuration JSON?

Crashillo replied:
Is this JSON config already available from de API? Our initial idea was fetch the site metadata from XHR call (because it should be stored in DB), instead having a json config file in the code

eltercero committed:
Replacing hardcoded routing methods with router-link

eltercero committed:
Typo fix.

eltercero commented:
Things left to be done in other issues:

- [ ] Check if we can change axios for d3.
- [ ] Add slots to the Index > Table > TableRow components
- [ ] Remove duplicated code in ContractsIndex and TendersIndex if TendersIndex remains.
- [ ] Move contracts and tendersData to props in router

@ferblape if everything else is correct I'll merge it

eltercero commented:
@ferblape is the review finished? can I merge it?

ferblape commented:
Yep 🚀

Crashillo commented:
Wait wait wait

Crashillo commented:
Blat the Kid, the fastest merge on this side of the Mississipi

Crashillo commented on app/javascript/gobierto_dashboards/webapp/components/TableRow.vue:
```diff
@@ -0,0 +1,58 @@
+        } else {
+          _self.formattedItem[column.field] = _self.item[column.field]
+        };
+      });
```
Using `self = this` is unnecessary in ES6. Convert the function inside the forEach into an arrow function to keep the context `this`

Crashillo replied:
Besides, you should apply destructuring always you have somethign like `obj.property` since that could be return error when column is null or undefined

eltercero replied:
Done! Thanks, way better.

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Nav.vue:
```diff
@@ -0,0 +1,74 @@
+        tag="li"
+        class="dashboards-home-nav--tab"
+      >
+        <i class="fas fa-chart-bar" />
+        <i class="far fa-chart-bar" />
```
Why are all icons twice?

eltercero replied:
I think it's related to the styling, I copied it from the investments navigation bar.

Crashillo replied:
Try single ones to notice if there's something wrong. If not, remove duplicates

eltercero replied:
The icon won't appear with just one:
 
![Captura de pantalla 2020-05-14 a las 10 28 55](https://user-images.githubusercontent.com/545235/81911543-c1c57900-95cd-11ea-931f-80669ae3eab8.png)

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Nav.vue:
```diff
@@ -0,0 +1,74 @@
+      labelTenders: I18n.t("gobierto_dashboards.dashboards.contracts.tenders")
+    }
+  },
+  routesToNavBarMapping: {
+    'summary': 0,
+    'contracts_index': 1,
+    'contracts_show': 1,
+    'tenders_index': 2,
+    'tenders_show': 2
+  },
```
Perhaps that should be defined at router level, above all is shared all around the site

eltercero replied:
I did it this way because I think the router may not need to know anything about the presentation (those are only the tab index for the Nav component).

Crashillo commented on app/javascript/gobierto_dashboards/webapp/lib/get_remote_data.js:
```diff
@@ -0,0 +1,13 @@
+import * as d3 from 'd3'
```
Use axios instead

eltercero replied:
noted! https://github.com/PopulateTools/gobierto/pull/3053#issuecomment-628052353

Crashillo commented on app/javascript/gobierto_dashboards/webapp/lib/utils.js:
```diff
@@ -0,0 +1,10 @@
+export function formatCurrency(amount) {
```
this file could be move to the vue-filters (`lib/shared`) or even that, check there if there's some common function you may reuse for your purpose

eltercero replied:
yeah, this is already removed in #3057 (https://github.com/PopulateTools/gobierto/pull/3057/commits/faab468f320a668fd523658f9efa1f274ff50393) because it will use an already created filter.

eltercero committed:
Removing d3 since it won't be used for now.

eltercero committed:
Changing function to arrow function

To avoid using _self as this. Also (not related), adding destructuring.

eltercero commented:
@Crashillo I think everything is either done or postponed (:P), can you check again?

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -1,3 +1,118 @@
+      const router = new VueRouter({
+        mode: "history",
+        routes: [
+          { path: "/dashboards/contratos", component: Home,
+            children: [
+              { path: "resumen", name: "summary", component: Summary },
+              { path: "contratos", name: "contracts_index", component: ContractsIndex },
+              { path: "contratos/:id", name: "contracts_show", component: ContractsShow },
+              { path: "licitaciones", name: "tenders_index", component: TendersIndex },
+              { path: "licitaciones/:id", name: "tenders_show", component: TendersShow },
+            ]
```
@eltercero Don't you tell me you gonna add the props here?

eltercero replied:
Yes, but I'll do it in the latest issue on development to avoid conflicts in three separate branches https://github.com/PopulateTools/gobierto/pull/3053#discussion_r424976360
